[Parse] [Sema] Update confusables diagnostic to mention the character names as well

theblixguy opened the pull request:
This is a follow up to #33047.

When we encounter confusable characters, we emit a diagnostic note to let the user know about it. This PR improves the diagnostic to include the names of the confusable and expected characters. This is quite helpful in situations where both the confusable and expected characters look identical to the user and the character names help provide more clarity and information.

For example, previously we would say:

```swift
print("A"); print("B") // identifier ';' contains possibly confused characters; did you mean to use ';'?
```
(and notice that both characters _look identical_)

and now we say:

```swift
print("A"); print("B") // identifier ';' (Greek Question Mark) looks similar to ';' (Semicolon); did you mean ';' (Semicolon)?
```

The character names are only present when there is a single code point (like above, which is a tailored diagnostic that I have added). Otherwise, they are not shown as there could be various code points and it could make the diagnostic a lot longer.

The Rust language also has a similar [diagnostic](https://github.com/rust-lang/rust/blob/master/src/test/ui/parser/unicode-chars.stderr) which mentions the character names, from which I have taken inspiration.

theblixguy committed:
[Parser] Update 'Confusables.def' file to include confusable and base character names

theblixguy committed:
[Parser] Add a new utility method to return the names of the confusable and base characters for a given confusable codepoint

theblixguy committed:
[Parser] Update diagnostic for confusable character during lexing to mention confusable and base character names

theblixguy committed:
[Sema] If there is just a single confusable character, emit a tailored diagnostic that also mentions the character names

theblixguy committed:
[Diagnostics] Add new diagnostic messages to the localization file

theblixguy committed:
[Test] Update confusables test

rintaro commented:
Whoa, I just noticed that we have a generator of `include/swift/Parse/Confusables.def`.

https://github.com/apple/swift/blob/master/utils/generate_confusables.py
https://github.com/apple/swift/blob/master/utils/UnicodeData/confusables.txt

Could you update them and generate the file from it?

theblixguy commented:
@rintaro Yeah, I noticed it after I manually updated the `def` file! It would be a little tricky though, because some names are formatted differently. For example, I use `Back Slash` instead of `Solidus` for expected character name for clarity. Would you prefer I change them in the `txt` file or perform ad-hoc substations in the script file? Or would you prefer I keep the mappings as-is?

rintaro commented:
I don't think it's a good idea to update `confusables.txt` because it's from [unicode.org](https://www.unicode.org/Public/security/9.0.0/confusables.txt). 
Ad-hoc substitutions sounds good.

Lastest `confusables.txt` is probably https://www.unicode.org/Public/security/13.0.0/confusables.txt

rintaro commented:
Could you add some warning text in the generated file like:
```
////////////////////////////////////////////////////////////////////////////////
// WARNING: This file is manually generated from
// utils/UnicodeData/confusables.txt and should not be directly modified.
// Run utils/generate_harness/generate_confusables.py to regenerate this file.
////////////////////////////////////////////////////////////////////////////////
```

theblixguy commented:
Yeah, I will add that!

theblixguy committed:
[Utils] Update unicode confusables txt file and update script to regenerate confusables def file

theblixguy committed:
[Parse] Regenerate 'Confusables.def' using updated script

theblixguy commented:
@rintaro I have updated the script and regenerated the `def` file.

rintaro commented on lib/Parse/Confusables.cpp:
```diff
@@ -14,9 +14,22 @@
+  default:
+    return std::make_pair("", "");
+  }
+}
```
Nit pick: newline

rintaro commented on include/swift/Parse/Confusables.def:
```diff
@@ -2,127 +2,134 @@
+// WARNING: This file is manually generated from
+// utils/UnicodeData/confusables.txt and should not be directly modified.
+// Run utils/generate_confusables.py to regenerate this file.
+////////////////////////////////////////////////////////////////////////////////
```
I prefer this under the `//===--- ... ---===//` line.

rintaro commented on utils/generate_confusables.py:
```diff
@@ -79,19 +82,30 @@ def main(args=sys.argv):
+////////////////////////////////////////////////////////////////////////////////
 '''
         f.write(header)
         f.write("//===----------------------------------------------------")
         f.write("------------------===//\n\n")
```
I wonder why this line (and the above) is not included in `header` 🤔

theblixguy replied:
I have moved it inside `header` now.

rintaro commented on utils/generate_confusables.py:
```diff
@@ -79,19 +82,30 @@ def main(args=sys.argv):
+            newExpectedName = expectedName
+            if expectedName in mappings:
+                newExpectedName = mappings[expectedName]
+            f.write("CONFUSABLE(" + confused + ", " + '"' + confusedName + '"' + ", " + expected + ", " + '"' + newExpectedName + '"' + ")\n")
```
This probably fails `validation-test/Python/python_lint.swift`. Try `@swift-ci Python lint`

theblixguy replied:
I have fixed it locally, but will run CI to double-check!

rintaro commented on utils/UnicodeData/confusables.txt:
```diff
@@ -1,11 +1,11 @@
-# Unicode Security Mechanisms for UTR #39
-# Version: 9.0.0
+# Unicode Security Mechanisms for UTS #39
+# Version: 13.0.0
```
I will confirm if we can simply update this file in our repo.

theblixguy replied:
Okay!

theblixguy committed:
[Utils] Adjust generate_confusables script based on review feedback

Fix a mistake with name mapping. Updated header comment. Fix a couple of linting issues.

theblixguy committed:
[Parse] Regenerate 'Confusables.def' file once again after script changes

theblixguy committed:
[Parse] Add the newline after end of 'getConfusableAndBaseCodepointNames' method

theblixguy commented:
@swift-ci please smoke test

theblixguy commented:
@swift-ci python lint

rintaro commented:
@theblixguy We can update `confusables.txt`. Could you fix the failing test case so we can merge?

theblixguy committed:
[Test] Update diagnostic message in 'Syntax/Parser/diags.swift'

theblixguy commented:
@rintaro Great! I have updated the test.

theblixguy commented:
@swift-ci please smoke test

rintaro commented:
Windows: [![Build Status](https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-windows/5581/badge/icon)](https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-windows/5581/)

rintaro commented:
@swift-ci Please smoke test Windows

theblixguy commented:
@swift-ci please Python lint
